fix: regression: sort day by fileCreatedAt again

midzelis opened the pull request:
## Description

Recent change introduced caused regression - days were sorted by localDateTime instead of fileCreatedAt. This change fixes that regression. 

- adds fileCreateAt, fileCreatedAtTimeZone, dayGroup to TimeBucketAssetResponseDto
- refactors, renames date utilities in timeline-util.ts
- moves TimeBucketAssetResponseDto to TImelineAsset routine to timeline-util, since thats where other TimelineAsset transforms are also performed
- updates tests
- addresses TODO in getTimelineBucket - returns a short date-only version of the timeBucket
- typos

docs and new tests in diff PR


Fixes # [(issue)](https://github.com/immich-app/immich/issues/18623#issuecomment-2913825835)

midzelis committed:
fix: regression: sort day by fileCreatedAt again

mertalev commented on web/src/lib/stores/assets-store.svelte.ts:
```diff
@@ -465,33 +468,9 @@ export class AssetBucket {
-          : null,
-        thumbhash: bucketAssets.thumbhash[i],
-      };
+      const timelineAsset = toTimelineAssetFromTimeBucketResponse(bucketAssets, i);
```
This code is not reusable anyway and it's more overhead in a hot area, so I think it's an unnecessary change.

mertalev commented on server/src/dtos/time-bucket.dto.ts:
```diff
@@ -78,7 +78,12 @@ export class TimeBucketAssetResponseDto {
+
+  fileCreatedAt!: string[];
+
+  @ApiProperty({ type: 'array', items: { type: 'string', nullable: true } })
+  fileCreatedAtTimeZone!: string[];
```
Why do we need the time zone? I thought the plan was fileCreatedAt + local day?

midzelis replied:
As it turns out, local day is a little problematic. The asset-store actually has multiple data sources, 1) loading time-buckets from the /api - but also, 2) websocket updates, and 3) manually updates calling `addAsset` or `updateAssets`, or `updateAssetOperation` (a result from some ui-actions) 

Since loading assets or updating assets in this way do not use a `TimeBucketAssetResponseDto` - we can't really use dayGroup. In fact, we could probably just drop dayGroup from the TimeBucketAssetResponseDto - kept it in for now, since @zackpollard specifically requested it. Not sure if based on this new info would you change your mind? 

The new technique is to send timeZone with fileCreatedDate. With fileCreatedDate always being UTC, it can be used directly for sorting. With the TZ info, we can derive localDateTime from the UTC by converting the UTC back to  tz-relative date, and then dropping the TZ. (TimelineAsset dosn't store tz in its plainDateFormats anyways) this derived localDateTime is used for grouping days/months - and for the day's formatted date string (title/heading).

mertalev replied:
The `timeZone` can be in one of many formats, including longer ones like `America/New_York`. It's a little unfortunate if we need to include it. It definitely shouldn't be in addition to the `dayGroup`.

midzelis committed:
lint

midzelis committed:
Merge remote-tracking branch 'origin/main' into localDateTime_regression

midzelis committed:
e2e test

midzelis committed:
inline function

midzelis committed:
e2e

alextran1502 commented:
Tested and confirmed the ordering issue. has been fixed

midzelis committed:
Address comments. Drop dayGroup and timezone in favor of localOffsetMinutes

midzelis committed:
Merge remote-tracking branch 'origin/main' into localDateTime_regression

midzelis committed:
lint and some api-doc

midzelis committed:
lint, more api-doc

midzelis committed:
format

midzelis committed:
Merge remote-tracking branch 'origin/main' into localDateTime_regression

midzelis committed:
Move minutes to fractional hours

midzelis committed:
make sql

alextran1502 commented:
@midzelis Can you help resolve the conflict of the PR?

@mertalev @zackpollard can you help with a review and stamp if they look good from your end?

mertalev commented on web/src/lib/stores/assets-store.svelte.ts:
```diff
@@ -478,9 +480,13 @@ export class AssetBucket {
         livePhotoVideoId: bucketAssets.livePhotoVideoId[i],
-        localDateTime: fromLocalDateTimeToObject(bucketAssets.localDateTime[i]),
+        localDateTime: fromISODateTimeWithOffsetToObject(
+          bucketAssets.fileCreatedAt[i],
+          bucketAssets.localOffsetHours[i],
+        ),
+        fileCreatedAt: fromISODateTimeUTCToObject(bucketAssets.fileCreatedAt[i]),
         ownerId: bucketAssets.ownerId[i],
-        people,
+        people: [],
```
Unnecessary memory usage

mertalev commented on web/src/lib/utils/timeline-util.ts:
```diff
@@ -5,17 +5,65 @@ import { AssetTypeEnum, type AssetResponseDto } from '@immich/sdk';
+export const fromISODateTimeToObject = (isoDateTime: string, timeZone: string): TimelinePlainDateTime =>
+  (fromISODateTime(isoDateTime, timeZone) as DateTime<true>).toObject();
+
+// used for AssetResponseDto.localDateTime, amoungst others
```
amongst

mertalev commented on web/src/lib/stores/assets-store.svelte.ts:
```diff
@@ -478,9 +480,13 @@ export class AssetBucket {
         isVideo: !bucketAssets.isImage[i],
         livePhotoVideoId: bucketAssets.livePhotoVideoId[i],
-        localDateTime: fromLocalDateTimeToObject(bucketAssets.localDateTime[i]),
+        localDateTime: fromISODateTimeWithOffsetToObject(
+          bucketAssets.fileCreatedAt[i],
+          bucketAssets.localOffsetHours[i],
+        ),
+        fileCreatedAt: fromISODateTimeUTCToObject(bucketAssets.fileCreatedAt[i]),
```
This seems to parse fileCreatedAt into DateTime twice per asset

midzelis committed:
Resolve merge conflicts between localDateTime_regression and main

- Resolved conflict in assets-store.svelte.ts: kept HEAD refactor and added missing isSelectingAllAssets export
- Resolved conflict in asset-factory.ts: kept HEAD imports for TimelineAsset and fromISODateTimeUTCToObject

midzelis committed:
merge/conflict

midzelis committed:
merge fallout, review comments

midzelis committed:
spelling

mertalev commented on server/src/queries/asset.repository.sql:
```diff
@@ -262,9 +262,16 @@ with
+      )::real / 3600 as "localOffsetHours",
       "assets"."ownerId",
       "assets"."status",
+      "assets"."fileCreatedAt",
```
DateTime can interpret it as UTC though, right?

midzelis replied:
Yes, it should. I'll update

zackpollard reviewed (approved):
LGTM!

danieldietzler commented on server/src/dtos/time-bucket.dto.ts:
```diff
@@ -5,72 +5,202 @@ import { AssetOrder, AssetVisibility } from 'src/enum';
 import { Optional, ValidateAssetVisibility, ValidateBoolean, ValidateUUID } from 'src/validation';
 
 export class TimeBucketDto {
+  @ApiProperty({
+    type: 'string',
+    format: 'uuid',
+    required: false,
+    description: 'Filter assets by specific user ID',
+  })
```
It's a bit annoying that you repeat a lot of stuff here that's already in the other decorators. I thought you can just add a comment above the field and it'll automatically turn it into a description? So like

```ts
// Filter assets by specific user Id
@ValidateUUID({ optional: true })
userId?: string;
```

midzelis committed:
drop offset from returned date

midzelis committed:
move description into decorator where possible, regen api
